### PR TITLE
Water - Mackenzie 

### PR DESCRIPTION
**Pull request by mvlofthus**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | I actually wrote custom controller methods. I used the model for validation. In the model, I sorted and ordered sets of works. 
Describe how you approached testing that model method. What edge cases did you come up with? | I used a controller method.  For the model validation for works for example, I made sure to test case sensitivity, duplicates in same category, and duplicates in different categories.    
What are `session` and `flash`? What is the difference between them? | Flash stores an error message, and can be shown for a full or partial request cycle (if flash or flash.now respectively).  Session can be used to store a user until a user logs themselves out or clears their cookies or exits the window, it is stored for the whole of a user's session or interaction with the app.  
What was one thing that you gained more clarity on through this assignment? |  I gained more clarity on indirect relationships and on how css can be used through bootstrap.  I also gained more clarity on db seeds, how to make them and use them, and yml files for test databases (and how those can be accessed with indirect relationships).  
What is the Heroku URL of your deployed application? | https://macklo-media-ranker.herokuapp.com/works

# Assignment Submission: Media Ranker

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection

Prompt | Response
--- | ---
What was a custom model method you wrote? What was it responsible for doing? | 
Describe how you approached testing that model method. What edge cases did you come up with? | 
What are `session` and `flash`? What is the difference between them? | 
What was one thing that you gained more clarity on through this assignment? | 
What is the Heroku URL of your deployed application? |

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -0,0 +1,74 @@
+class UsersController < ApplicationController
+  def index
+    @users = User.all
+  end
+
+  def show
+    @user = User.find_by(id: params[:id])
+
+    if @user.nil?
+      redirect_to users_path, status: :temporary_redirect
```
Comment (kaidamasaki) [inline app/controllers/users_controller.rb:10]: Nice use of a temporary redirect!

File: app/controllers/votes_controller.rb [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+  def create
+    user = User.find_by(id: session[:user_id])
+    work = Work.find_by(id: params[:work_id])
+
+    session[:return_to] = request.referer
+
+    if user.nil?
+      flash[:warning] = "You must be logged in to vote on works"
+      redirect_back(fallback_location: root_path)
```
Comment (kaidamasaki) [inline app/controllers/votes_controller.rb:11]: Nice use of `redirect_back`!

File: app/views/works/edit.html.erb [hunk 3]
```diff
@@ -0,0 +1,11 @@
+<h2>Edit this <%= @work.category.titleize %></h2>
```
Comment (kaidamasaki) [inline app/views/works/edit.html.erb:1]: `titleize`!

**Review by kaidamasaki — COMMENTED** [review]
Good job!  You didn't finish implementing everything but what you did have looked great!

I really liked your different statuses on your `flash` being `success`, `warning` and `danger`.  

Going forward I'd like you to brush up on controller filters and custom model methods.  It's a little late in the project but if you can manage to get some work done on them in bEtsy that would be great (either implementation or tests).  If not I'd suggest just reading up on them and practicing a little over break week.

# Media Ranker





## Functional Requirements: Manual Testing

| Criteria | yes/no |
| --- | --- |
| **Before logging in** | --
| 1. On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | ✔️
| 2. Can go into a work's show page | ✔️
| 3. Verify unable to vote on a work, and get a flash message | ✔️
| 4. Can edit this work successfully, and get a flash message | ✔️
| 5. Can go to "View all media" page and see three lists of works, sorted by vote | ✔️
| 6. Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | ✔️
| 7. Can create a new work successfully. Note the URL for this work's show page | ✔️
| 8. Can delete this work successfully |✔️
| 9. Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | ✔️
| 10. Verify that the "View all users" page lists no users | Seeded users.
| **Log in** | --
| 11. Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | ✔️
| 12. Your username is listed in "View all users" page | ✔️
| 13. Verify that number of votes determines the Media Spotlight | ✔️
| 14. Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | ✔️?
| 15. Voting on the same work twice produces an error and flash message, and there is no extra vote | ✔️
| **Log out** | --
| 16. Logging out showed a flash message and changed the UI | ✔️
| 17. Logging in as a new user creates a new user | ✔️
| 18. Logging in as an already existing user has a specific flash message | ✔️

## Major Learning Goals/Code Review

| Criteria | yes/no |
| --- | --- |
| 1. Sees the full development cycle including deployment, and the app is deployed to Heroku | ✔️
| 2. Practices full-stack development and fulfilling story requirements: the styling, look, and feel of the app is similar to the original Media Ranker | 
| 3. Practices git with at least 25 small commits and meaningful commit messages | ✔️

### Previous Rails learning, Building Complex Model Logic, DRYing up Rails Code

| Criteria | yes/no |
| --- | --- |
| 4. Routes file uses `resources` for works | ✔️
| 5. Routes file shows intention in limiting routes for voting, log-in functionality, and users | ✔️
| 6. The homepage view, all media view, and new works view use semantic HTML | ✔️
| 7. The homepage view, all media view, and new works view use partials when appropriate | ✔️
| 8. The model for media (likely named `work.rb`) `has_many` votes | ✔️
| 9. The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | No custom model methods.
| 10. Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | No controller filters.
| 11. Some controller, likely the `WorksController`, has a controller filter for finding a work | No controller filters.
| 12. The `WorksController` uses strong params | ✔️
| 13. The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | ✔️

### Testing Rails Apps

| Criteria | yes/no |
| --- | --- |
| 14. There are valid fixtures files used for users, votes, and works | ✔️
| 15. User model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 16. Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | ✔️
| 17. Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 18. Work model has tests with a section on all business logic methods in the model, including their edge cases | No custom model methods.

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 14+ in Functional Requirements: Manual Testing && 14+ in Code Review | 
| Yellow (Approaches Standards) | 12+ in Functional Requirements: Manual Testing && 11+ in Code Review, or the instructor judges that this project needs special attention | ✔️
| Red (Not at Standard) | 0-10 in Code Review or 0-11 in Functional Reqs, or assignment is breaking/doesn’t run with less than 5 minutes of debugging, or the instructor judges that this project needs special attention | 





## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Elegant/Clever | ✅
| Descriptive/Readable | ✅
| Concise | Controller filters would have helped DRY up your code.
| Logical/Organized | ✅
